File: 백준/Silver/1629. 곱셈/곱셈.java
import java.io.*;
import java.util.*;

public class Main {

    public static void main(String[] args) throws IOException {

        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        br.close();

        long A = Integer.parseInt(st.nextToken());
        long B = Integer.parseInt(st.nextToken());
        long C = Integer.parseInt(st.nextToken());

        System.out.println(solution(A, B, C));
    }
    static long solution(long a, long b, long c) {

        // a의 0제곱이면 1
        if(b == 0) {
            return 1;
        }
        // b가 홀수 제곱이면 lower(b/2) * a값을 해줘야함
        if(b % 2 == 1) {
            return a * solution(a, b - 1, c) % c;
        }
        long temp = solution(a, b / 2, c) % c;
        return temp * temp % c;
    }
}

File: 백준/Silver/1629. 곱셈/곱셈.py
import sys
input = sys.stdin.readline

a, b, c = map(int, input().split())
    
def solution(a: int, b: int, c: int):
    if b == 0: # 재귀의 끝(b가 0이 되면 1 리턴)
        return 1
    if b % 2 == 1: # 홀수일 경우 a^b = a * a^(b-1)
        return a * solution(a, b-1, c) % c
    temp = solution(a, b//2, c) % c # 짝수일 경우 a^(b/2) * a^(b/2)
    return temp * temp % c # a^(b/2) * a^(b/2)

print(solution(a,b,c))